src/server/clientiface.cpp

```diff
@@ -432,12 +432,17 @@ void RemoteClient::SentBlock(v3s16 p)
 
 void RemoteClient::SetBlockNotSent(v3s16 p)
 {
-
 	// remove the block from sending and sent sets,
 	// and reset the scan loop if found
 	if (m_blocks_sending.erase(p) + m_blocks_sent.erase(p) > 0) {
 		m_nothing_to_send_pause_timer = 0;
-		m_nearest_unsent_d = std::min(m_nearest_unsent_d, m_last_center.getDistanceFrom(p));
+		// Reset to the maximum cube size that will
+		// guarantee that this block will the scanned again.
+		// Using m_last_center is OK, as a change in center
+		// will reset m_nearest_unsent_d to 0 anyway (set getNextBlocks).
+		p -= m_last_center;
+		m_nearest_unsent_d = std::min(m_nearest_unsent_d,
+				(s16)std::max(std::max(std::abs(p.X), std::abs(p.Y)), std::abs(p.Z)));
```

Comment:
This is the main fix. Instead of the euclidean distance, we need to find the cube side from which we need restart the scanning. This is always <= `getDistanceFrom`. (If there's a nicer way to express this, I am all ears.)